yb-voyager/src/migassessment/gather.go

```diff
@@ -0,0 +1,566 @@
+			displayName = displayName[:maxDisplayNameLen-3] + "..."
+		}
+		tracker.displayNames[node.nodeName] = displayName
+
+		if len(displayName) > tracker.maxNameLen {
+			tracker.maxNameLen = len(displayName)
+		}
+
+		// Initialize with pending stage
+		tracker.statuses[node.nodeName] = &NodeProgress{
+			NodeName: node.nodeName,
+			Stage:    "Pending...",
+		}
+	}
+
+	return tracker
+}
+
+func (pt *progressTracker) update(progress NodeProgress) {
+	pt.mutex.Lock()
+	defer pt.mutex.Unlock()
+
+	// Update only the stage, preserve the formatted & truncated display name
+	if existing := pt.statuses[progress.NodeName]; existing != nil {
+		existing.Stage = progress.Stage
+	}
+
+	// Print/update display
+	pt.printAll()
+}
+
+func (pt *progressTracker) printAll() {
+	if !pt.initialized {
+		// First time: print all lines
+		fmt.Println() // Blank line
+		for _, nodeName := range pt.nodes {
+			status := pt.statuses[nodeName]
+			pt.printSingleLine(*status)
+		}
+		pt.initialized = true
+	} else {
+		// Move cursor up to the first line and reprint all
+		// Move up by number of nodes
+		fmt.Printf("\033[%dA", len(pt.nodes))
+
+		// Reprint all lines
+		for _, nodeName := range pt.nodes {
+			status := pt.statuses[nodeName]
+			pt.printSingleLine(*status)
+		}
+	}
+}
+
+func (pt *progressTracker) printSingleLine(progress NodeProgress) {
+	// Determine icon based on stage
+	var statusIcon string
+	switch progress.Stage {
+	case "Complete":
+		statusIcon = "✓"
+	case "Failed":
+		statusIcon = "⚠"
+	default:
+		statusIcon = "⏳"
+	}
+
+	// Look up display name from the map (already truncated in newProgressTracker)
+	displayName := pt.displayNames[progress.NodeName]
+
+	// Clear line and print with fixed-width column for name (for alignment)
+	// \r returns to start, \033[K clears to end of line
+	// Using %-*s for left-alignment with dynamic width based on longest (truncated) name
+	fmt.Printf("\r\033[K  %s %-*s %s\n", statusIcon, pt.maxNameLen+1, displayName+":", progress.Stage)
+}
+
+// GatherAssessmentMetadataFromPG collects metadata from PostgreSQL primary and replicas.
+// Accepts the validated replicas to collect from, and returns the list of failed replica names
+// (for reporting partial multi-node assessment).
+// Collection is performed in parallel for better performance.
+func GatherAssessmentMetadataFromPG(
+	source *srcdb.Source,
+	validatedReplicas []srcdb.ReplicaEndpoint,
+	assessmentMetadataDir string,
+	pgssEnabled bool,
+	iopsInterval int64,
+) (failedReplicaNodes []string, err error) {
+	scriptPath, err := findGatherMetadataScriptPath(POSTGRESQL)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build list of all nodes to collect from (primary + replicas)
+	nodes := []collectionNode{
+		{
+			nodeName:      "primary",
+			connectionUri: source.DB().GetConnectionUriWithoutPassword(),
+			isPrimary:     true,
+			replica:       nil, // nil for primary
+		},
+	}
+
+	for i := range validatedReplicas {
```

Comment:
if there are no replicas, it might be better to follow the old path (where we list our all the steps one by one, because "primary"/"replica" terminology will be confusing to the user.